sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/RecordReaderIterator.scala

```diff
@@ -52,4 +55,18 @@ class RecordReaderIterator[T](rowReader: RecordReader[_, T]) extends Iterator[T]
     havePair = false
     rowReader.getCurrentValue
   }
+
+  override def close(): Unit = {
+    if (rowReader != null) {
+      try {
+        // Close the reader and release it. Note: it's very important that we don't close the
```

Comment:
This comment is copied from `NewHadoopRdd`, which contains similar defensive programming.